### ISSUE TITLE
New option: ignore_empty_lines

### ISSUE THREAD

**Issue by KenPowers** [editorconfig/editorconfig#238]
Did a quick search and didn't find anything being discussed. Some editors (and linters) support ignoring empty or whitespace-only lines when detecting trailing whitespace. Example, from Atom (despite Atom not yet supporting `trim_trailing_whitespace`):

![image](https://cloud.githubusercontent.com/assets/174864/10978344/67ae6e1a-83c4-11e5-88b5-e319c25536af.png)

I think it would be nice to be able to configure this behavior through EditorConfig as some projects remove all extraneous whitespace whereas some leave whitespace-only lines.

**Comment by cdutson** [comment]
:+1: I'd like this as well. When writing code, I add blank lines to separate logical blocks of code. When the file get's processed, it removes all those lines, and those logical separations as a result.

**Comment by jedmao** [comment]
@cdutson you say "it removes all those lines". What is "it" in this context?

**Comment by florianb** [comment]
@cdutson, @knpwrs for which plugins does this apply? And is this mentioning empty lines within code or at the end of the file?

**Comment by evan-king** [comment]
I like to keep whitespace-only lines loaded with the same level of indentation as the block they're in, which turns me off to trimming trailing whitespace.  If I could preserve that, I'd be happy to purge any pointless spaces after code elsewhere.

**Comment by florianb** [comment]
Hmm.. i am sorry, but i think i don't get it.

If you are talking about empty lines within your code, they shouldn't be affected by any rule yet. So in order to discuss this issue i think i need a more precise description what this is about.

**Comment by jedmao** [comment]
I think they are talking about indentation on empty lines being trimmed off when saving a file. They want to preserve indentation on empty lines so they can click or arrow-key navigate the file without going back to character zero.

**Comment by florianb** [comment]
Ahh.. thanks @jedmao.. this makes sense..

**Comment by jedmao** [comment]
I believe this issue might be better filed in the specific editor in question, as EditorConfig is only really meant to enable the built-in editor setting or trigger it.

Definitely curious which editors we're talking about. I think the most recent vscode version exposed a new editor setting for this, so it just needs to be implemented.

Basically, EditorConfig should say whether to do the trim it not, but your actual editor settings should determine HOW it's applied.

**Comment by florianb** [comment]
Hmm.. this issue applies at least for atom-editorconfig. And the definition says trim trailing whitespaces, which is exactly what it does.

If this is no intended behavior i'd rather suggest making the rule more explicit than changing it for one editor.

**Comment by evan-king** [comment]
Isn't the whole point of editorconfig to make formatting choices declaratively consistent in an editor-agnostic way and thus ease development in heterogenous environments?  It kind of defeats the purpose if different editors are going to differently interpret the rules.  What's being requested as I understand it (trim whitespace only after non-whitespace characters) is a feature tweak that to my knowledge isn't natively implemented in any editor.

The resulting solution for me has traditionally been "don't trim whitespace" which isn't popular in communities/workplaces that want to enforce style/formatting.  I would like to see a compromise for whitespace-only lines introduced, though I acknowledge the effort to implement is not small, especially for editors that cannot gain the feature via plugin.

**Comment by jedmao** [comment]
@evan-king, I totally get your argument that it defeats the purpose if different editors are going to differently interpret the rules, but you have to also consider the fact that different editors have different philosophies about how certain things should be.

For example, it's not even possible for you to save a file in vscode with mixed line endings, because they made it impossible. Whether you agree with this implementation or not is irrelevant. The point is that it would be unwise to then ensure any/all EditorConfig plugins behave exactly the same way, because then it would mean introducing, in this case, vscode philosophies into other editors, which have completely different takes on the issue of mixed line endings.

Having said all this, I believe it's not the goal of EditorConfig to create a 100% consistent experience between editors, but to "align" with those editors in such a way that they continue to work the way they were designed. Also, to "leverage" the built-in editor features as much as possible. This also ensures the users of those editors are not surprised by EditorConfig doing some hacky thing to fight against the editor implementations.

Does that make sense? I'm proposing that plugins just tell the respective editors to trim trailing whitespace and let the editor decide "how" that is accomplished. If there is a setting in your editor to ignore empty lines, that's great. Enable it if you want that. But if the editor doesn't have that setting? Then it doesn't have that setting. That's what I'm saying.

Related: https://github.com/editorconfig/editorconfig-vscode/issues/57

**Comment by evan-king** [comment]
Fair point, and overall I agree.  I think, however, that if EditorConfig is going to ask for whitespace to be preserved on blank lines, then editors which won't respect that should also be told not to trim trailing whitespace at all.

In other words, degrade features non-destructively.  In a homogenous environment you wouldn't have trimming at all, and in a heterogenous environment the code would converge to trimmed only where desired via the participation of editors which fully accommodate the spec.

The alternative opens up an edit war (or code review noise over trivial issues) between people trying to preserve the whitespace that .editorconfig tells them they can have and the people obliviously using editors which blast it away.

**Comment by jedmao** [comment]
What I'm pitching is that we don't introduce this new `ignore_empty_lines` rule. There doesn't have to be editor wars, because if the editor doesn't work the way you want then you just file the issue with the editor until it does or contribute a PR yourself (if it's that important to you). I just don't think the value proposition is strong enough here to warrant the introduction of a new rule.

**Comment by florianb** [comment]
I am a little bit confused. The editorconfig-page states at first

> EditorConfig helps developers define and maintain consistent coding styles between different editors and IDEs

Personally i really expect any editorconfig-rule to be applied exactly the same way in every different tool. Except a rule cannot be implemented because of limitiations within implementing environment. Which the should be just a not or partly implemented property.

But this is just something i see as a foundation - i am open for a discussion.

Creating a new rule is probably overkill - refining the `trim_trailing_whitespaces`probably isn't? In my opinion it is legitimately to ask if this behavior shouldn't be part of the editorconfig-specification in any way (this could also mean, that the rule-definition would suggest a preferred way of implementation).

**Comment by ffes** [comment]
> Personally i really expect any editorconfig-rule to be applied exactly the same way in every different tool. Except a rule cannot be implemented because of limitiations within implementing environment.

I agree with you.

If people **really** want/need this option I think adding the value `ignore_empty_lines` to `trim_trailing_whitespaces` seems like the best solution. I have no need for this.

And I know that implementing this in the Notepad++ plugin will not be trivial. Now it just calls a builtin function. If this feature needs to be implemented it will have to go through all the lines and determine what to do. I think this goes for most editors.

**Comment by jedmao** [comment]
@florianb, the mission statement of EditorConfig to maintain consistent coding styles between different editors is an excellent one, but that's not exactly true today and it may never be true in the future. There are various reasons for this, one of which is the fact that certain rules are a bit difficult to implement in a particular editor. Also, each editor tends to have a completely separate developer writing the plugin, which means we can't all be 100% in sync.

What I'm suggesting is that we merely leverage built-in editor features where we can and don't fight the editor to do something it wasn't designed to do. In the case of vscode, it's actually **_impossible_** to implement the `end_of_line` setting consistently with other editors in that vscode is not designed to allow you to save a file with mixed line endings. @ffes has just stated that implementing the proposed `ignore_empty_lines` rule or setting would "not be trivial." I think this is a sign that we're going the wrong direction and off the golden path.

In conclusion, I think we can still achieve the mission statement of EditorConfig (i.e., maintaining consistent coding styles between different editors) without fighting against the respective editors as long as we understand it's actually not 100% possible. We can do our best, but we shouldn't have to go out of our way and do something crazy complicated to support a feature if the editor itself doesn't provide the functionality we need.

**Comment by florianb** [comment]
@jedmao thanks for sharing your very interesting point of view!

So you're saying -- as far as i understood you -- that we aim for standardizing editor-properties where we know it is possible they can be implemented. And we do not consider functionality where it's proven it can't be implemented.

If this is what you think how the editorconfig-standardization should work i think i have a complete different understanding of it. Probably there are more opinions on that?

I think it is impossible to define rules where it's assured they can be implemented in the intended editors. And the current situation of editorconfig proves that.

If this is how editorconfig works, where is the list of the "respective" editors? I don't know of any list of aimed editors for editoroconfig. But even if there is one (and i would be really thankful if i would get a hint where to find it) this approach would - in my eyes - include at least the following risks:

1) Editors come and go - fostering a list of target-browsers would add an immense amount of work to have a valid list where the standardization board could rely on.
2) I state it is impossible to find a subset of functionality where it is possible to define its intentional behavior in such a standardized way that its implementation is consistent over the different editors.

In addition to that i agree standardizing on a current well known intersection of editor-features is a way to start doing thing. Now editorconfig has grown and gets mature. It is legitimate to consider widening the functionality, raising the standard.

You say the golden path is sticking with things where its well known how this might be implemented and where we can assure tools are 100% in sync. I don't know if is understand this. Let's take `max_line_length` as an example. Until i implemented it into atom-editorconfig the tools couldn't never have been 100% in sync. But let's say we're in sync because we just count that this property is somehow implemented. I guess we're still not in sync again, since its behavior is implemented different across the several editors. And there is even no warranty that the mention of supporting editors is in any way complete or exhaustive.

I would like to see aiming for an ideal definition of properties, which serves the users of editors the best of we can think of. Of course we can't assure that these definitions are properly and exhaustively implemented in all these hundreds of editors. We could just leave this business to all the users, plugin-developers and developers of those editors. At the end its the user enforcing the implementation.

So instead of dropping a discussion just because it seems hard to implement we should think about if the benefit of a rule may be worth the process of proper definition (not implementation).

To be honest i don't know if this applies to the proposed `ignore_empty_lines` but thats another topic which just should be fairly discussed.

**Comment by jedmao** [comment]
@florianb, you said that I said...
> The golden path is sticking with things where its well known how this might be implemented and where we can assure tools are 100% in sync.

I don't think I said that anywhere. What [I said](https://github.com/editorconfig/editorconfig/issues/238#issuecomment-268595974) is quite the opposite:

> I believe it's not the goal of EditorConfig to create a 100% consistent experience between editors, but to "align" with those editors in such a way that they continue to work the way they were designed.

I [also said](https://github.com/editorconfig/editorconfig/issues/238#issuecomment-268882412) that, "[super complicated implementations are] a sign that we're going the wrong direction and off the golden path." That's the only time I mentioned the golden path.

The "list of respective editors" I'm talking about are [the ones listed on the download section of the front page of EditorConfig.org](http://editorconfig.org/#download) (i.e., ones that support EditorConfig "in some way").

I also am not suggesting we drop something because it's hard to implement. I'm only suggesting that if it's so hard, it's probably because we're doing it wrong and there might be a more simple way to do it that aligns with the design goals of the editor in question.

As for guidelines for plugin authors, I've created https://github.com/editorconfig/editorconfig/issues/295 and we can continue the discussion there.

**Comment by florianb** [comment]
Thanks for clarification @jedmao.

**Comment by cesarsouza** [comment]
I am also interested in this issue. I suppose it could be worth to align configuration settings between at least VS, VS Code, VS for Mac and MonoDevelop. I do not have much details to share at this time though, but considering recent contributions to at least one open-source repository here at GitHub (https://github.com/migueldeicaza/TensorFlowSharp) it is possible that those editors can offer different options regarding keeping indentation for empty/whitespace lines. It would be nice if we could use .editorconfig files to sync settings at least across those editors.

If I understand correctly how .editorconfig works, it can also be that this setting can be implemented as extensions (plugins?) particular to each of those editors, is it right? If yes, I can also suggest those changes directly to https://github.com/dotnet/roslyn/issues/18342 or https://github.com/dotnet/roslyn/pull/15065.

**Comment by andrewrothman** [comment]
I think having this option would be awesome. Enabling this would greatly improve my workflow, as I often find myself having to add code to functions which are already indented and have spaces between separate chunks of functionality. I find myself having to press tab many times a day to properly indent my code (I use VS Code, and collaborators often use the built-in auto-formatter which removes all trailing lines including whitespace only lines).

The editorconfig project is a godsend, but please consider adding support for this feature!

**Comment by mxxcon** [comment]
I also want this feature for my projects and to have consistently formatted files. I do not want 'space-filled blank lines' between my code blocks and this option would ensure that.

**Comment by tajmone** [comment]
I was about to open an issue for this feature request, and was glad to see that other users preceded me.

Many editors allow preserving indentation in empty line, and some languages come with native editors that enforce this. The rationale for indenting empty lines following indented code is to simplify additions by having the right indentation already in place.

For example (spaces represented with `•`):

```purebasic
If 5 > 0
••
  Debug "True!"
••
EndIf
```

I'm experiencing problems with EditorConfig validation of projects that contain sources with indented empty lines, forcing me to drop validation of trailing whitespace in the sources by setting `.editorconfig` to:

    trim_trailing_whitespace = unset

### EditorConfig for CI Validation

I'm adding to all my repos `.editorconfig` settings so that I can validate every commit and PR on Travis CI via [EClint]  (a Node.js CLI validator for EditorConfig). Using EditorConfig to validate project contents is a very effective way to preserve code styles consistency in repositories.

I'm starting to realize that there are subtle differences on how EditorConfig affects editing and validation. For the former, it serves mainly as a guiding tool to inform the editor on how to handle the code inside a project, whereas in the latter case it serves a stricter role of enforcer/validator of the actual sources.

Often discussions of EditorConfig focus only on the editor side, forgetting that the validation part is just as important — after all, you need to ensure that these settings are being honoured in the actual code contributed by third parties.

### Further Thoughts on This Feature

Enabling EditorConfig settings to allow empty lines indentation is definitely a needed feature. Here are some further thoughts on the issue:

- Should this lead to blindly ignore any whitespace-only lines? or
- Should validators check that the indentation of the empty line matches that of the preceding line?

Ideally, the latter would be more inline with the feature intentions, where an indentation beyond the preceding code should be considered an error. This might be harder to implement in validators (whereas editors only need to know if they should indent empty lines or not, for they already have the means to handle indentations correctly).

The former is definitely easier to implement for plug-ins developers.

### Against Naming It `ignore_empty_lines`

I'm also assuming that with this new feature being introduced, `indent_size` and `indent_style` would still apply to empty lines, therefore I think that this option shouldn't be called `ignore_empty_lines` for it would misleading if empty lines indentation need to satisfy the `indent_size` and `indent_style` criteria.

A more fitting name might be `allow_empty_lines_indentation`, which delivers more concisely the idea that the source code is allowed to preserve indentation in empty lines (but still needs to honour `indent_size` and `indent_style`, if they apply).




[EClint]: https://www.npmjs.com/package/eclint

**Comment by jedmao** [comment]
Without language-specific AST knowledge, there's no way to know which lines are considered empty or not. Consider a multi-line string where some of the lines are empty. You wouldn't want EditorConfig touching those lines.

As EditorConfig remains agnostic of any particular language, this feature is unsupportable.

I suggest you do `trim_trailing_whitespace = unset` or just `false`.

**Comment by tajmone** [comment]
> Without language-specific AST knowledge, there's no way to know which lines are considered empty or not.

Technically speaking, a line containing whitespace is not an empty line. The idea of allowing indentation on empty lines is to allow EditorConfig to consider the white space as indentation (if it meets the criteria of correct `indent_style` and `indent_size`); from the EditorConfig point of view, the line is "empty" because it contains only indentation (followed by an EOL).

I'm not quite sure which of the above proposals you see as unachievable without an AST, but I guess you're referring to the correct indentation level.
Anyhow, it would be still a great improvement if it was possible to allow _indentation only_ lines (aka indented empty lines) without sacrificing the checks for correct `indent_style` and `indent_size`.

> Consider a multi-line string where some of the lines are empty. You wouldn't want EditorConfig touching those lines.

Again, there's a difference on how these options should be interpreted by editors and code validators. I wouldn't want a validator to touch anything (although tools like EClint are quite good at fixing problems).

An EditorConfig validation tool wouldn't need to treat such a string as a special case, but only need to know if indented empty lines should be allowed or not — the editor should anyhow preserve indentation styles for empty lines too. Whether such strings would be rendered with an indented empty line or not is irrelevant from a validation point of view, for both a (truly) empty line and one consisting of indentation only would be fine in this context.

> As EditorConfig remains agnostic of any particular language, this feature is unsupportable.

Remaining language-agnostic is one thing, but EditorConfig can't be (and isn't) _coding patterns agnostic_. I fail to see how indentation of (otherwise) empty lines falls under specific languages, it's a recurrent pattern in various languages and markup syntaxes (not to mention autogenerated files of various types), all of which might be desirable to be validated against `.editorconfig` settings in collaborative projects.

I think this feature might be an example of EditorConfig options that could be devised with validation in mind, rather than editing. If empty lines indentation can be set, we shouldn't really worry about how editors are going to handle them, for we can assume that if they are commonly found in a language there'll already be editors able to handle it.

It's the _lack of this option_ that affects trailing whitespace validation of sources that might contain indented empty lines. My understanding is that people desiring this feature are asking that EditorConfig tools would treat the whitespace on (otherwise) empty lines as indentation, instead of trailing space.

After all, since these lines have no other contents there isn't a _before_ and _after_, so the very concept of _leading_ and _trailing_ are somewhat obfuscated.

One might legitimately ask why the spaces in these lines is currently treated as trailing whitespace — indeed, since it occurs at the beginning of the line it should be considered indentation. There's no trailing whitespace on empty lines, for there are no contents to be followed by whitespace!

When one thinks of the term leading- and trailing-whitespace it comes natural to think of how whitespace-trimming functions work, and how these two concepts are relative to the presence of some non-whitespace content on the string (or line) in question.

If EditorConfig would have considered the initial whitespace of these lines as indentation, we wouldn't even be discussing this issue because these lines wouldn't break code validation.

What are the arguments for considering space-only lines as trailing spaces?

> I suggest you do trim_trailing_whitespace = unset or just false.

I'm already doing this, for I have no other choice.

**Comment by cxw42** [comment]
I support this feature request.  I have contributed to open-source projects that leave whitespace-only lines within code blocks in Perl POD documentation.  Some POD renderers have historically broken code blocks at each empty line.  So instead of:

```
foo
                    <--- empty line here, with no whitespace
bar
```
you would get:
```
foo
```
```
bar
```
if you didn't leave trailing whitespace on the line after `foo`.

I suggest adding a new value `trim_trailing_whitespace = nonspace` to trim trailing whitespace only on lines that have one non-space character.  This is similar to https://github.com/editorconfig/editorconfig/issues/238#issuecomment-268761638 , but I think we shouldn't use the word "empty" since a whitespace-only line isn't empty, but looks empty in many editors.

Re. AST, if this property is defined to only operate line-by-line, then `/\H/` would be all the test we would need to determine whether to trim trailing whitespace on a line.  That might have to be implemented by hand in some editors, but it requires only programmatic access to each line independently.  This setting is orthogonal to how much whitespace is added to any particular line.

**Comment by jedmao** [comment]
What if the line in question has 20 spaces, even though it’s at a single indentation level? Do we trim down to 4 spaces (if that’s the setting)?

This is a trick question, because there’s no way to distinguish between indentation and alignment without AST. Furthermore, there’s no way to detect if a particular line (in isolation) is a multi-line string or not.

There’s only so much we can do from an agnostic standpoint and as I’ve said time and time again, the job of EditorConfig is not to have these custom line by line operations provided by extensions. If the editor provides such an operation, we can leverage it, but outside of that, we should back off.

**Comment by evan-king** [comment]
Non sequitur.  How many spaces do you leave on non-empty lines when not trimming their trailing whitespace for language X?  No one's asking for code reformatting, but rather for what content mutating is _already_ performed be a little more considerate so that the feature is more broadly usable instead of another way for trivial developer preferences to trod on each other's toolchains.

**Comment by jedmao** [comment]
I think you’re missing the point. If it requires a custom line by line implementation by means of an extension then it’s outside the goals of EditorConfig, which is intended to assist built-in editor functionality via configuration.

**Comment by cxw42** [comment]
@jedmao I'm not sure I understand why EditorConfig is limited to built-in editor functionality.  For example, Vim actually does not have trim-trailing-whitespace built-in (that I know of).  The plugin implements `trim_trailing_whitespace` by doing a custom regex replace every time the file is about to be saved:

https://github.com/editorconfig/editorconfig-vim/blob/53c56fcf358ca36b00b9cbf7dd70642c8922fdd0/plugin/editorconfig.vim#L434-L437

You have been around much longer than I have, but my impression during my time with EditorConfig has been that EditorConfig specifies behaviour, independently of how plugins implement that behaviour.

**Comment by jedmao** [comment]
I apologize for the confusion. I perhaps took it a bit too far. Please refer to the [Plugin Guidelines](https://github.com/editorconfig/editorconfig/wiki/Plugin-Guidelines) for implementation advice.

I would argue, however, that Vim's implementation has the potential to be destructive in a multi-line string scenario, just as I'm sure other extensions are doing as well.

I agree to a point that EC should be focused on the what and not the how, but I also think we need to remind extension developers or [guide them](https://github.com/editorconfig/editorconfig/wiki/Plugin-Guidelines) as to how to do it in such a way that it's non-destructive. That's one motivation for my comments. Another is that various editors have their own sort of philosophy that go with them (e.g., when you change your newline setting in vscode it changes it for every line). I wouldn't suggest that extensions do that sort of thing, but if the editor is going to do it, let it do its thing and stay out of the editor's kitchen.

**Comment by jedmao** [comment]
For the record, I'm guilty of [the same destructive implementation](https://github.com/editorconfig/editorconfig-vscode/blob/08d184d50a74f383367a8272680cde208a17b7ad/src/transformations/TrimTrailingWhitespace.ts#L91). I just wish there were a more editor-supported path that didn't require this non-AST modification!

**Comment by cxw42** [comment]
@jedmao Thanks for the clarification!  Your point about multi-line strings is a reasonable one.  I guess I just haven't tripped over that one myself (I trim trailing WS aggressively in all my own projects :) ).

**Comment by tajmone** [comment]
The more I think about this issue and the more I realize that the problem lies with considering the whitespace in _indented empty lines_ as trailing whitespace. There is no trailing whitespace in empty lines, only indentation.

The concept of _leading-_ and _trailing-_ whitespace is clearly in reference to actual contents being there. Since there is no representation of a line-beginning, but end-of-lines are indeed represented by the EOL sequence (which is also considered belonging to the whitespace category), when a line contains only whitespaces we're dealing with and indentation of the end-of-line — not trailing whitespace after the line beginning.

This can be easily grasped by thinking of the problem in terms of RegExs, where the line beginning anchor `^` doesn't actually represent a real character in the text, whereas the end of line anchor `$` matches either `LF`/`CRLF` (or EOF).

If empty lines with only whitespace where already interpreted this way, we wouldn't even need be asking for this new feature.

This can (and ultimately will) have strong implications on markup syntaxes, especially in non-Western languages which rely on special whitespace chars or include RTL directionality (often switching from one to the other in mixed texts).

The goal of "syntax agnosis" should be shaped around real-case scenarios, and keep into account also spoken languages, not only programming languages — the term "syntax agnostic" would actually be better, for it would include non only programming languages, for editors are used for editing in different contexts, including text. Besides, regardless of the programming language, people  comment their code in their native language, so it's not unusual to find multi-line comments in RTL languages.

If the EditorConfig gets the leading- vs trailing- concept wrong, all sorts of problems will raise with RTL languages, for when a source file switches language direction it can get tricky to distinguish instantiation from trailing spaces — unless you take the EOL as a reference to distinguish what comes first and what follows contents (or the lack thereof).

**Comment by florianb** [comment]
I know a bunch of applications where `trim_trailing_whitespace` is used to remove unnecessary indentations of empty lines. I also use a linter for my daily work which errors that case out.

I fear a change of `trim_trailing_whitespace`'s behaviour would just swap things to other peoples issues.

And i would say this feature is context sensitive, to make it right we need to know the semantics. EditorConfig core privides a context free ruleset.

But if we could provide different categories or rulesets of features which operate f.e. context free or context sensitive (or headless), we would enable the next generation of tools (modern editors, linters, etc) to make fully use of their contextual knowledge like ASTs?

**Comment by jedmao** [comment]
I think the Prettier project is more equipped to deal with these context sensitive scenarios via their [plugins API](https://github.com/prettier/prettier/blob/master/docs/plugins.md), as they are AST-aware.

**Comment by tajmone** [comment]
> I know a bunch of applications where `trim_trailing_whitespace` is used to remove unnecessary indentations of empty lines. I also use a linter for my daily work which errors that case out.

Modern editors tend to have dedicated settings for preserving indentation on empty lines following indented code (Sublime Text does), so they might be able to interpret EditorConfig directives according to context.

Possibly, if this feature is named `preserve_indented_empty_lines`, editors will be able to contextualize that in this case whitespace only lines should be interpreted as indentation instead of trailing whitespace.

But then, the sad think about programming standards is that they were (and are) rarely designed with an outlook on all spoken languages in mind. History shows that most standards are US-centric, from the ASCII standard onward (the "American" in ASCII doesn't certainly include Latin America and the letters of the alphabets of the languages spoken there).

Just think of the "human readable" concept and how it's applied to markdown and other lightweight markup syntaxes, which heavily rely on characters like `` ` `` and `~`, which are mostly absent on non-US/UK keyboards (unlike the `$` which is found on every keyboard, for some reason). Human readability, as concept, ultimately translates to having to constantly <kbd>Alt</kbd>+<kbd>&lt;code&gt;</kbd> on the num-pad to access back-ticks and tildes, if you happen to live in a non-anglophone country.

Rarely standards can afford to be truly unbiased, unless they do make a strong effort to reach out to experts on i18n problems — and, from my experience, native speakers who feel their needs constantly ignored and stamped-on in standards design rarely make the first move to point the problem out, being so used as they are in being treated as "edge cases".

The problem is that these biases rarely stand still, instead they tend to propagate with a snow-ball effect, with every next standard inheriting the good as well as the bad of the standards that preceded. Human nature is prone to prefer smooth paths and quick results to impervious and unexplored ways.

Whenever you mention RTL languages (like Arabic) in some editor's forum, you usually get replied that "edge cases" can't and won't be covered. The truth is that if you look at statistics, native speakers of LTR Latin-based languages are the minority world-wide. If you look at "spoken languages" you get a biased result, for many countries which have a non-Latin native language happen to also speak a second language due to colonial heritage.

Having officially entered the era of Unicode, I think it's time to be open minded regarding standards in general, and I personally find it amusing that today there are still editors unable to support LTR languages like Arabic, Persian and Hebrew inside comments. I mean ... Lua let's you name identifiers in full Unicode. NP++, for example, does a fairly good job at support non Latin languages (better than most commercial editors).

Ultimately, to avoid these pitfalls and achieve true "language agnosis" organizations only need to include in their decision making process native speakers of various languages, so they can get feedback on the implications and backlashes of their decisions.

**Comment by florianb** [comment]
I was not saying that your use case does not exist. I said the opposite use case exists, too. Which makes changing current behaviour a bad decision.

The LTR-argument is good, but `trim_trailing_whitespace` is in my opinion already reading direction agnostic.

And i'd like to join @jedmao point, the rule would need context. And a context-aware tool could do the job.
And i would also take the pressure from this issue to evolve EditorConfig in a backwards compatible way for modern use cases.

**Comment by tajmone** [comment]
Fine then. I only suggest that you make it clear in the documentation what is meant by whitespace — the term can vary from context to context, whereas in common-use editors is usually understood as meaning only Ascii spaces and tabs, while in Unicode it covers a variety of spaces entities (including vertical tabs, various widths spaces, and locale-specific spaces for some non-Latin languages, etc.).

**Comment by jedmao** [comment]
Yeah, we're trying to [lock some of this down](https://github.com/jedmao/editorconfig-ini/blob/master/src/ini.pest#L25).

**Comment by tajmone** [comment]
@florianb:

> And i would say this feature is context sensitive, to make it right we need to know the semantics.

I've been thinking about what you said, and checked real case scenarios where indented-only lines occur. My conclusion is that this feature doesn't require an AST nor knowledge of the language semantics.

Usually indentation-only lines are found in source files of language IDEs or editors that by default preserve indentation in empty lines; in these cases you'll only find such lines between equally indented elements. E.g.

```
    print "xxx"
    <-- ( indented "empty" line )
    print "xxx"
```

In cases where the empty line occurs between a shift in indentation level, the empty line usually indent to the larger indentation of the two elements:

```
if x=1
    <-- ( indented "empty" line )
    Then ...
```

In any case, this type of indent empty line is always the result of how a given editor works, so it's automatically handled by the IDE/Editor linter.
For these reasons, I'd say that those who do need this feature are usually dealing with auto-generated indentations on empty lines, which are managed by their IDE/Editor.

My guess is that this feature will not serve to guide an editor on how to handle indentation of empty lines (although some general purpose editors allow this option, e.g. Sublime Text, but even then they auto-manage the indentation size by their own algorithm); instead it will be used more for code-style consistency validation (e.g. via Travis CI).

Currently, in all my repositories where I enforce code styles via EditorConfig, and use EClint for sources validations via Travis CI, I have to disable trailing spaces checks for sources that are known to contain indented empty lines, and have to use a separate script and validate them via grep (which somehow weakens the purpose of using EditorConfig).

In such cases (and I can't think of many others) what would be expected from the validator is simply to skip trailing whitespace checks on lines that contain only indentation (i.e. as defined in `.editorconfig` in terms of either spaces or tabs, and width). The width of similar empty lines indentation would still have to obey the general rules.

**Comment by cxw42** [comment]
@tajmone thanks for collecting the data, and for the use-case explanation.  

> skip trailing whitespace checks on lines that contain only indentation (i.e. as defined in .editorconfig in terms of either spaces or tabs, 

I can see how this would work, and I think it's very straightforward in the validator case.  It also seems easy in the editor case, since it treats each line independently.

>and width

Would you please explain this a bit more?  I thought checks for a specific width required context or language knowledge.

Thanks!

**Comment by tajmone** [comment]
> Would you please explain this a bit more? I thought checks for a specific width required context or language knowledge.

I see two possible context-free solutions to this, which can overlap each other without conflict, or just adopt one of the two:

1. Check that the indented empty line(s) don't exceed the indentation of the previous or following non empty lines (whichever has the wider indentation).
2. Check that the indentation of empty line is a valid multiple of `indent_size` (as usual) — i.e. if `indent_size=1` any width is OK, if it's `=2` then it must be a multiple of 2, and so on.

The former doesn't check if the width is a valid multiple of `indent_size`.

Having just the latter would already make a huge practical difference, but hopefully the two are not mutually exclusive.

**Comment by tajmone** [comment]
As a practical real-case scenario I face every day: I work with a language that ships with its native IDE (its linter indents empty line when you reformat the source file), but I mostly work with Sublime Text, for which I created my custom syntax for that language (which I've set to don't preserve indentation on empty lines).

In a collaborative effort on GitHub we soon realized that my commits would introduce huge diffs dues to WS differences in empty lines (all other aspects are controlled via EditorConfig and Tavis CI, except `trim_trailing_whitespace` due to this missing feature).

Unfortunately, Sublime Text is not able to enforce those empty line indentations when reformatting code, it only supplies them while you type.

So the solution was to freely edit the sources with any editor, but always reformat the code with the native IDE before committing, so no WS or indentation (and formatting) difference creep into the commits.

This is the natural use case, for it mostly applies to languages which have a native IDE, which is always available to its end users, therefore the solution is practical.

Beside that, there might be the occasional extra space or tab accidentally left over by forgetting to reformat, or by headless manual addition — in any case, these would (and should be caught) and usually don't upset the whole code base.

In all other cases (i.e. the language doesn't specify to indent empty lines, nor it enforces it in its native IDE), the best solution is to always avoid empty indented lines, and let EClint treat them as `trim_trailing_whitespace` errors. In my case this was not possible due to the IDE being closed source (at the time being).

So really, this feature request is important for these edge cases, for it's currently an impediment to using `trim_trailing_whitespace` at all — and we've experienced so many problems with extra WS garbage commits that in the end it was becoming hard to sift through changes, and we had to enforce a custom script to catch all trailing WSs on lines which contained at least one non WS char (needless to say, it required using grep with a greedy RegEx that is quite slow).

In all other cases, EditorConfig and EClint are a true blessing shielding the repos from wild code-styles and editors fighting over control of the code by their own custom settings reformatting whole source files.
